registry.ts

```diff
@@ -154,6 +154,65 @@ async function unpkgVersions(name: string): Promise<string[]> {
   return m.map((x) => x[1]);
 }
 
+interface PackageInfo {
+  parts: string[];
+  scope: string;
+  packageName: string;
+  version: string;
+}
+function defaultParts(that: RegistryUrl): PackageInfo {
+  const parts = that.url.split("/");
+  const [packageName, version] = parts[4].split("@");
+  if (parts[3] === undefined) {
+    throw new Error(`Package scope not found in ${that.url}`);
+  }
+  if (packageName === undefined) {
+    throw new Error(`Package name not found in ${that.url}`);
+  }
+  if (version === undefined) {
+    throw new Error(`Unable to find version in ${that.url}`);
+  }
+  return {
+    scope: parts[3],
+    packageName,
+    version,
+    parts,
+  };
+}
+
+export class UnpkgScope implements RegistryUrl {
+  url: string;
+
+  parts(): PackageInfo {
+    return defaultParts(this);
+  }
+
+  constructor(url: string) {
+    this.url = url;
+  }
+
+  async all(): Promise<string[]> {
+    const { scope, packageName } = this.parts();
+    return await unpkgVersions(`${scope}/${packageName}`);
+  }
+
+  at(version: string): RegistryUrl {
+    const { parts, packageName } = this.parts();
+    parts[4] = `${packageName}@${version}`;
+    return new UnpkgScope(parts.join("/"));
```

Comment:
To decrease some boilerplate perhaps:
```
function defaultScopeAt(that: RegistryUrl): string {
   const { parts, packageName } = defaultParts(this);
   parts[4] = `${packageName}@${version}`;
   return parts.join("/");
}
```
could be pulled out or something?

Reply:
Looks good!
I consider whether to merge `defaultScopeAt` into `defaultAt` like this:

```ts
export function defaultAt(
  that: RegistryUrl,
  version: string,
  hasScope = false,
): string {
  if (!hasScope) return that.url.replace(/@(.*?)(\/|$)/, `@${version}/`);

  const { parts, packageName } = defaultParts(that);
  parts[4] = `${packageName}@${version}`;
  return parts.join("/");
}
```

but this refactoring merely makes registry.ts little shorter, so I'll use the code you suggested.